### ISSUE TITLE
[compile time] AOT Autograd is taking log time in tracing

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#125977]
### 🐛 Describe the bug

profile - https://www.internalfb.com/intern/scuba/query/?dataset=pyperf_experimental%2Fon_demand&drillstate=%7B%22purposes%22%3A%5B%5D%2C%22end%22%3A%22now%22%2C%22start%22%3A%22-30+days%22%2C%22filterMode%22%3A%22DEFAULT%22%2C%22metrik_view_params%22%3A%7B%7D%2C%22modifiers%22%3A%5B%5D%2C%22sampleCols%22%3A%5B%5D%2C%22cols%22%3A%5B%22namespace_id%22%2C%22namespace_process_id%22%5D%2C%22derivedCols%22%3A%5B%5D%2C%22mappedCols%22%3A%5B%5D%2C%22enumCols%22%3A%5B%5D%2C%22return_remainder%22%3Afalse%2C%22should_pivot%22%3Afalse%2C%22is_timeseries%22%3Afalse%2C%22hideEmptyColumns%22%3Afalse%2C%22dimensions%22%3A%5B%22process%22%2C%22service_tags_app%22%5D%2C%22timezone%22%3A%22America%2FLos_Angeles%22%2C%22compare%22%3A%22none%22%2C%22samplingRatio%22%3A%221%22%2C%22metric%22%3A%22count%22%2C%22aggregation_field%22%3A%22async_stack_complete%22%2C%22top%22%3A%2210000%22%2C%22enrich%22%3A%22None%22%2C%22api%22%3A%22SQL%22%2C%22aggregateList%22%3A%5B%5D%2C%22param_dimensions%22%3A%5B%7B%22dim%22%3A%22py_async_stack%22%2C%22op%22%3A%22all%22%2C%22param%22%3A%220%22%2C%22anchor%22%3A%220%22%7D%5D%2C%22order%22%3A%22weight%22%2C%22order_desc%22%3Atrue%2C%22constraints%22%3A%5B%5B%7B%22column%22%3A%22sample_tags%22%2C%22op%22%3A%22all%22%2C%22value%22%3A%5B%22%5B%5C%222024-05-10-15%3A30%3A503924488devgpu002.ash8.facebook.com%5C%22%5D%22%5D%7D%5D%5D%2C%22c_constraints%22%3A%5B%5B%5D%5D%2C%22b_constraints%22%3A%5B%5B%5D%5D%2C%22ignoreGroupByInComparison%22%3Afalse%7D&view=Icicle3&=&normalized=1715380267&pool=uber&icicle_state=%7B%22deltaMode%22%3A%22none%22%2C%22colorMode%22%3A%22total%22%2C%22valueMode%22%3A%22selection%22%2C%22formatMode%22%3A%22percentage%22%2C%22codeHubState%22%3A%7B%22aggregationField%22%3A%22%22%2C%22scubaParams%22%3A%22%22%2C%22symbolPath%22%3Anull%7D%2C%22showCodeProfileModal%22%3Afalse%2C%22colorPatterns%22%3A%5B%5D%7D

xref - https://fb.workplace.com/groups/1075192433118967/permalink/1425136311457909/

AOT Autograd is taking large amount of time (get breakdown).

### Versions

N/A

**Comment by ezyang** [comment]
This one high prio?

**Comment by anijain2305** [comment]
Making it high priority, mostly to get an owner to take a look and check for low hanging fruits if any.

**Comment by bdhirsh** [comment]
I ran the repro locally, switching out the backend for `aot_eager_decomp_partition` to only include non-inductor bits.

I found:

(1) the total time was 378s. 131s of that was coming from `_extract_graph_with_inputs_outputs`.

(2) I noticed that @aorenste recently had a PR to speed up this function due to inefficient lookups of graph inputs ([PR](https://github.com/pytorch/pytorch/pull/125937)), that hadn't made it into fbcode yet.

When I patch in that PR, the total drops to 245s. Still not great, but a >50% speedup

**Comment by bdhirsh** [comment]
Trying to think more about where the low-hanging fruit in `FunctionalTensor.__torch_dispatch__` is:

(1) 6s: I tried making `return_and_correct_aliasing` a no-op ([here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/functional_tensor.py#L511)), and it shaved ~6 seconds off the repro (251s -> 245s). It's possible that @jbschlosser 's changes here https://github.com/pytorch/pytorch/pull/126552 will help (the fn shouldn't have to do a lot of work, just returnining the right argument and swapping storages for view ops).

(2) 21s: removing custom size dispatch on `FunctionalTensor` [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/functional_tensor.py#L139). Right now, every call to `.shape()` goes through __torch_dispatch__ of functional tensor, then `FunctionalTensorWrapper`, then `FakeTensor`. I don't... think this is actually necessary? When I changed it to `None`, I get a drop from 245s -> 224s.

(3) ~4s. When I print out any other metadata calls that are getting plumbed through FunctionalTensorMode [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/functional_tensor.py#L457), I see many calls to `prim.device.default`. Looking at the stacktrace from one of them, it looks like every time we construct a fresh `FunctionalStorageWrapper` [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalStorageImpl.cpp#L95), the inner tensor is a `FakeTensor` so we go through the full set of mode dispatch. I got the 4s just from hardcoding that value to `Device(DeviceType::CUDA)`.

I'm not sure if (3) is worth the work immediately, but (2) definitely seems worth attempting to fix (PR incoming)

**Comment by Chillee** [comment]
How do I look at the SVG for this?

**Comment by laithsakka** [comment]
> How do I look at the SVG for this?

curious is there any thing specfic about SVG that makes it easier to navigate than strobelight 
graph profile ( https://fburl.com/scuba/pyperf_experimental/on_demand/n5umzh5x)
, or icicle profile ( https://fburl.com/scuba/pyperf_experimental/on_demand/ghc8lpjr) ?

also:
you can get a SVG like visualization by clicking on visualize from the graph profiler view
https://fburl.com/scuba/pyperf_experimental/on_demand/04q2l2de

<img width="1037" alt="Screenshot 2024-05-22 at 10 10 44 AM" src="https://github.com/pytorch/pytorch/assets/13970752/e241d580-a0b2-4f32-8ed1-301e0686ed2e">

**Comment by Chillee** [comment]
@laithsakka There's a lot of things I often want to know about the data that I find hard to figure out from other views, such as

Graph View
- what functions are calling other functions? (which graph view doesn't give)
- What is the overall function-call hierarchy?

Icicle View
- The icicle view does not handle situations such as "A calls C" and "B calls C" - how much total time is spent in C?
- It also handles recursion pretty poorly.

The SVG view in strobelight is pretty bizarre as well - it also doesn't seem like it handles recursive dependencies?

I prefer a view like this (i.e. what's gotten from `TORCH_COMPILE_CPROFILE`)
<img width="1258" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/b0812890-95c0-40aa-b723-cf176cbb08ee">

**Comment by masnesral** [comment]
From triage meeting: Brian still planning to get to this

**Comment by laithsakka** [comment]
> @laithsakka There's a lot of things I often want to know about the data that I find hard to figure out from other views, such as
@Chillee 
I see I see, btw Graph profiler does show callers of each functions once you click on it and percentage of calls coming from A vs B to C.
Also in Icicle View someone can select to **reverse the stack** in that case it will show C with A and B as children and so on, with weights showing percentage of calls coming from A vs B.

But that said I understand that different people have different preferences also.

**Comment by yf225** [comment]
monthly check: is this still high-priority? cc. @anijain2305 @bdhirsh

**Comment by bdhirsh** [comment]
I'm leaving this open as an umbrella issue (probably won't be fixed this month, but tracking as an important item to chip away at throughout H2). So given that, I think leaving it open as hi-pri seems reasonable?

a recent AOTAutograd compile time win was https://github.com/pytorch/pytorch/pull/131899, which sped up net compile time of `BertForMaskedLM` with `backend="aot_eager"` by 20% (waiting for more benchmarks here: https://github.com/pytorch/pytorch/actions/runs/10150417141)

**Comment by bdhirsh** [comment]
results from [fast-path-detach() benchmarking](https://github.com/pytorch/pytorch/actions/runs/10150417141) (I only ran "inductor default", for training and inference:

training compile times:
<img width="847" alt="image" src="https://github.com/user-attachments/assets/3f6275a9-f6f2-453b-a967-b8dbe6b05fe8">


inference compile times:
<img width="851" alt="image" src="https://github.com/user-attachments/assets/c10e4b33-3991-42e8-8874-4a917f9206b0">


Looks like the fastpath had a bigger impact on training. But ~15% compile time improvement for huggingface + training + inductor default is not bad :)

**Comment by bdhirsh** [comment]
fast-pathing `FunctionalTensor.shape` results for inductor / training defaults ([PR](https://github.com/pytorch/pytorch/pull/132084)):
<img width="847" alt="image" src="https://github.com/user-attachments/assets/41dc2bf9-8de7-43ee-8439-be0530727b88">

HF + training has the best compile-time geomean speedup at 15%

**Comment by bdhirsh** [comment]
status update:

(1) a few improvements to compile times in the last few weeks (see above)

(2) I probably won't have time to spend a lot more time on other improvements this half

(3) we still want to prioritize this overall though

A few other improvements that we are looking into (not sure if they will happen this half or later):

@tugsbayasgalan to investigate constructing the `ViewAndMutationMeta` without needing an entire round of tracing. This is a big chunk of work, but could improve the AOTDispatcher section of compile times by up to ~25-40% ish for inference (a bit less for training)

We also (eventually) want to look at making the dispatch order a bit more static, which should make the amount of time taken to proxy trace every individual op in the graph noticeably faster (design needed, maybe for next half)

**Comment by bdhirsh** [comment]
(oh fwiw, https://github.com/pytorch/pytorch/pull/132524 should also improve compile times from functionalization by some noticeable amount, although I spent a bit of time trying to benchmark and gave up)

**Comment by zou3519** [comment]
Don't know if I wrote this down anywhere else, but another way to chop compilation times is: AOTAutograd does a two pass thing. The first pass it computes some metadata (by running Functionalization with FakeTensor), the second pass it actually traces out a joint graph. If we could fuse these two it would save some time (usually there are 2x as many nodes in the backward as in the forward, so this could be up to 25% savings)

**Comment by shunting314** [comment]
From triage review: we will work on this in H1 2025.

**Comment by bdhirsh** [comment]
https://github.com/pytorch/pytorch/pull/145082 ended up being a nice win for AOT tracing (8% improvements on some microbenchmarks)

**Comment by shunting314** [comment]
From triage review meeting: There a few knows things we can do but no one sign up for them right now.
